### PR TITLE
cpu/sam0_common/periph_adc: add work around for errata 2.1.6

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This adds a delay between enabling the ADC and starting to sample on the SAMD5x MCUs when the internal bandgap reference is used.

image

### Testing procedure

Before, `adc_sample()` on one of the SAMD5x MCUs I have at hands got stuck when the internal bandgap reference was used by never leaving this loop:

```
0x00001660 in _sample (line=line@entry=0)
    at /home/marian.buschsieweke@ml-pa.loc/Repos/software/ML-PA-RIOT/cpu/sam0_common/periph/adc.c:489
489     while (!(dev->INTFLAG.reg & ADC_INTFLAG_RESRDY)) {}
```

Now, it works also on the "special" MCU that got stuck before:

```
> saul read 0
2024-11-14 22:14:45,179 # saul read 0
2024-11-14 22:14:45,180 # Reading from #0 (vol|SENSE_ANALOG)
2024-11-14 22:14:45,180 # Data:	          2122 
> saul read 1
2024-11-14 22:14:47,434 # saul read 1
2024-11-14 22:14:47,451 # Reading from #1 (cur|SENSE_ANALOG)
2024-11-14 22:14:47,452 # Data:	            18 
```

### Issues/PRs references

None

### CODE REVIEW

File: cpu/sam0_common/periph/adc.c [hunk 1]
```diff
@@ -239,6 +248,35 @@ static int _adc_configure(Adc *dev, adc_res_t res)
...
+     * > If CTRLA.ONDEMAND = 0: Add a delay of minimum 40 μs between the
+     * > enable of the ADC (CTRLA.ENABLE) and the start of the first conversion.
+     * > [...]
+     *
+     * We do so using ztimer if used anyway, or busy waiting otherwise.
+     */
+#  if MODULE_ZTIMER_USEC
+    ztimer_sleep(ZTIMER_USEC, 40);
+#  elif MODULE_ZTIMER_MSEC
+    ztimer_sleep(ZTIMER_USEC, 1);
```
[the remark this answers is not in the document]
Comment (maribu) [reply cpu/sam0_common/periph/adc.c:273]: Good catch, thx!

File: cpu/sam0_common/periph/adc.c [hunk 2]
```diff
@@ -239,6 +248,35 @@ static int _adc_configure(Adc *dev, adc_res_t res)
...
+#  else
+    /* let's rather be safe than sorry */
+    busy_wait_us(100);
```
[the remark this answers is not in the document]
Comment (maribu) [reply cpu/sam0_common/periph/adc.c:276]: The `busy_wait_us()` is just counting CPU cycles and not that precise. I make the comment above a bit more specific.

[the remark this answers is not in the document]
Comment (maribu) [reply cpu/sam0_common/periph/adc.c]: I also reduced it two twice the specified amount. That turns out to work reliably on my "bad MCU" affected by the bug and there should be enough safety margin.

**Review by dylad — APPROVED** [review]
ACK.
Tested on my `same54-xpro` with `tests/periph/adc` which is stuck if
`#define ADC_REF_DEFAULT                     ADC_REFCTRL_REFSEL_INTREF`
is used.
With this PR, it works as expected.

**Comment by maribu** [comment]
Thx :)

**Pull request merged**